### PR TITLE
AM-1085: bug fixed by implementing same behaviour in different scenarios

### PR DESCRIPTION
**Pull request by ashraf706**
### How to test

1. Create a B2B application and send a request to the authorize endpoint with a redirect uri such as [http://localhost:8092/domain/oauth/authorize?client_id=XXXredirect_uri=https%3A%2F%2Fgoogle.co.uk]
2. The response shouldn't redirect to the site mentioned in the redirect_uri
3. Now create a Web application with a redirect uri. 
4. Change the Web application to a B2B application.
5. Send a request to the authorize endpoint.
6. The outcome should be exactly as mentioned in 2. The response won't redirect and error msg should be same.

### CODE REVIEW

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/authorization/AuthorizationRequestFailureHandler.java [hunk 1]
```diff
@@ -289,4 +290,13 @@ private void doRedirect(RoutingContext context, String url) {
         cleanSession(context);
         context.response().putHeader(HttpHeaders.LOCATION, url).setStatusCode(302).end();
     }
+
+    /**
+     * Application, only with 'client_credentials' grant type cannot have
+     * `redirect_uri` in the request
+     */
+    private boolean invalidParamRedirectURI(Client client, String redirectURI) {
+        return client != null && client.getAuthorizedGrantTypes().contains(CLIENT_CREDENTIALS) &&
```
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/authorization/AuthorizationRequestFailureHandler.java:299]: **question**: Don't think it is useful but do we have to test also if `client.getAuthorizedGrantTypes` is null or not ?

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/authorization/AuthorizationRequestFailureHandler.java [hunk 2]
```diff
@@ -167,10 +168,10 @@ private void processOAuth2Exception(AuthorizationRequest authorizationRequest,
             authorizationRequest.setRedirectUri(defaultErrorURL);
         }
         // check if the redirect_uri request parameter is allowed
-        if (client != null
+        if ((client != null
                 && client.getRedirectUris() != null
                 && authorizationRequest.getRedirectUri() != null
-                && !client.getRedirectUris().contains(authorizationRequest.getRedirectUri())) {
+                && !client.getRedirectUris().contains(authorizationRequest.getRedirectUri())) || invalidParamRedirectURI(client, authorizationRequest.getRedirectUri())) {
```
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/authorization/AuthorizationRequestFailureHandler.java:174]: **nitpick**: maybe we can move the first condition in a method named `forbiddenRedirectUri` for readability :thinking: ?

` if (forbiddenRedirectUri(client, authorizationRequest) || invalidParamRedirectURI(client, authorizationRequest.getRedirectUri()))`

**Review by leleueri — APPROVED** [review]

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/authorization/AuthorizationRequestFailureHandler.java (hunk 1 above)
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/authorization/AuthorizationRequestFailureHandler.java]: done, null check added

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/authorization/AuthorizationRequestFailureHandler.java (hunk 2 above)
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/authorization/AuthorizationRequestFailureHandler.java]: done, refactored and moved to a method as suggested

**Pull request merged**